modal/image.py

```diff
@@ -275,17 +276,77 @@ class _Image(_Object, type_prefix="im"):
     force_build: bool
     inside_exceptions: List[Exception]
     _used_local_mounts: typing.FrozenSet[_Mount]  # used for mounts watching
+    _mounts: Sequence[_Mount]  # added as mounts on any container referencing the Image, see `def _mount_layers`
 
     def _initialize_from_empty(self):
         self.inside_exceptions = []
         self._used_local_mounts = frozenset()
+        self._mounts = ()
+        self.force_build = False
+
+    def _initialize_from_other(self, other: "_Image"):
+        # used by .clone()
+        self.inside_exceptions = other.inside_exceptions
+        self.force_build = other.force_build
+        self._used_local_mounts = other._used_local_mounts
+        self._mounts = other._mounts
 
     def _hydrate_metadata(self, message: Optional[Message]):
         env_image_id = config.get("image_id")
         if env_image_id == self.object_id:
             for exc in self.inside_exceptions:
                 raise exc
 
+    def _add_mount_layer_or_copy(self, mount: _Mount, copy: bool = False):
+        if copy:
+            return self.copy_mount(mount, remote_path="/")
+
+        base_image = self
+
+        async def _load(self2: "_Image", resolver: Resolver, existing_object_id: Optional[str]):
+            self2._hydrate_from_other(base_image)  # same image id as base image as long as it's lazy
+            self2._mounts = tuple(base_image._mounts) + (mount,)
+            self2._used_local_mounts = base_image._used_local_mounts | ({mount} if mount.is_local() else set())
+
+        return _Image._from_loader(_load, "ImageWithMounts()", deps=lambda: [base_image, mount])
+
+    @property
+    def _mount_layers(self) -> typing.Tuple[_Mount]:
+        """Non-evaluated mount layers on the image
+
+        When the image is used by a Modal container, these mounts need to be attached as well to
+        represent the full image content, as they haven't yet been represented as a layer in the
+        image.
+
+        When the image is used as a base image for a new layer (that is not itself a mount layer)
+        these mounts need to first be inserted as a copy operation (.copy_mount) into the image.
+        """
+        return self._mounts
+
+    def _assert_no_mount_layers(self):
+        if self._mount_layers:
+            raise InvalidError(
+                textwrap.dedent(
+                    """
+                It's recommended to run any `image.add_*` commands for adding local resources last
```

Comment:
Suggestions for how to make this wall of text easy to understand would be appreciated :D